In [0]:
# Simple CNN model for CIFAR-10
import numpy
from keras.datasets import cifar10
from keras.models import Sequential
from keras.layers import Dense
from keras.layers import Dropout
from keras.layers import Flatten
from keras.constraints import maxnorm
from keras.optimizers import SGD
from keras.layers.convolutional import Conv2D
from keras.layers.convolutional import MaxPooling2D
from keras.utils import np_utils
from keras import backend as K
K.set_image_dim_ordering('th')

In [2]:
# fix random seed for reproducibility
seed = 7
numpy.random.seed(seed)
# load data
(X_train, y_train), (X_test, y_test) = cifar10.load_data()

170500096/170498071 [==============================] - 9s 0us/step


In [0]:
# normalize inputs from 0-255 to 0.0-1.0
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
X_train = X_train / 255.0
X_test = X_test / 255.0


In [0]:
# one hot encode outputs
y_train = np_utils.to_categorical(y_train)
y_test = np_utils.to_categorical(y_test)
num_classes = y_test.shape[1]


In [5]:
# Create the model
model = Sequential()
model.add(Conv2D(32, (3, 3), input_shape=(3, 32, 32), padding='same', activation='relu', kernel_constraint=maxnorm(3)))
model.add(Dropout(0.2))
model.add(Conv2D(32, (3, 3), activation='relu', padding='same', kernel_constraint=maxnorm(3)))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Flatten())
model.add(Dense(512, activation='relu', kernel_constraint=maxnorm(3)))
model.add(Dropout(0.5))
model.add(Dense(num_classes, activation='softmax'))


Instructions for updating:
Colocations handled automatically by placer.
Instructions for updating:
Please use `rate` instead of `keep_prob`. Rate should be set to `rate = 1 - keep_prob`.


In [0]:
import keras
tbCallBack= keras.callbacks.TensorBoard(log_dir='./Graph', write_images=True)


In [13]:
# Compile model
epochs = 25
lrate = 0.01
decay = lrate/epochs
sgd = SGD(lr=lrate, momentum=0.9, decay=decay, nesterov=False)
model.compile(loss='categorical_crossentropy', optimizer=sgd, metrics=['accuracy'])
print(model.summary())


_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_1 (Conv2D)            (None, 32, 32, 32)        896       
_________________________________________________________________
dropout_1 (Dropout)          (None, 32, 32, 32)        0         
_________________________________________________________________
conv2d_2 (Conv2D)            (None, 32, 32, 32)        9248      
_________________________________________________________________
max_pooling2d_1 (MaxPooling2 (None, 32, 16, 16)        0         
_________________________________________________________________
flatten_1 (Flatten)          (None, 8192)              0         
_________________________________________________________________
dense_1 (Dense)              (None, 512)               4194816   
_________________________________________________________________
dropout_2 (Dropout)          (None, 512)               0         
__________

In [14]:
model.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=epochs, batch_size=32, callbacks=[tbCallBack])

Train on 50000 samples, validate on 10000 samples
Epoch 1/25
50000/50000 [==============================] - 10s 204us/step - loss: 0.8147 - acc: 0.7118 - val_loss: 0.9516 - val_acc: 0.6760
Epoch 2/25
50000/50000 [==============================] - 10s 203us/step - loss: 0.5716 - acc: 0.7994 - val_loss: 1.0123 - val_acc: 0.6788
Epoch 3/25
50000/50000 [==============================] - 10s 200us/step - loss: 0.4275 - acc: 0.8483 - val_loss: 1.0286 - val_acc: 0.6995
Epoch 4/25
50000/50000 [==============================] - 10s 204us/step - loss: 0.3353 - acc: 0.8827 - val_loss: 1.0235 - val_acc: 0.7011
Epoch 5/25
50000/50000 [==============================] - 10s 210us/step - loss: 0.2750 - acc: 0.9026 - val_loss: 1.0521 - val_acc: 0.7101
Epoch 6/25
50000/50000 [==============================] - 10s 201us/step - loss: 0.2326 - acc: 0.9174 - val_loss: 1.0931 - val_acc: 0.7118
Epoch 7/25
50000/50000 [==============================] - 10s 199us/step - loss: 0.1989 - acc: 0.9307 - val_loss: 1.

In [15]:
# Final evaluation of the model
scores = model.evaluate(X_test, y_test, verbose=0)
print("Accuracy: %.2f%%" % (scores[1]*100))
print("loss: %.2f%%" % (scores[0]*100))

Accuracy: 72.12%


In [0]:
model = Sequential()
model.add(Conv2D(32, (3, 3), input_shape=(3, 32, 32), padding='same', activation='relu', kernel_constraint=maxnorm(3)))
model.add(Dropout(0.2))
model.add(Conv2D(32, (3, 3), activation='relu', padding='same', kernel_constraint=maxnorm(3)))
model.add(MaxPooling2D(pool_size=(2, 2)))

model.add(Conv2D(64, (3, 3), activation='relu', padding='same', kernel_constraint=maxnorm(3)))
model.add(Dropout(0.2))
model.add(Conv2D(64, (3, 3), activation='relu', padding='same', kernel_constraint=maxnorm(3)))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Conv2D(128, (3, 3), activation='relu', padding='same', kernel_constraint=maxnorm(3)))
model.add(Dropout(0.2))
model.add(Conv2D(128, (3, 3), activation='relu', padding='same', kernel_constraint=maxnorm(3)))
model.add(MaxPooling2D(pool_size=(2, 2)))
# flattening the matrix into vector form
model.add(Flatten())
model.add(Dropout(0.2))
model.add(Dense(1024, activation='relu', kernel_constraint=maxnorm(3)))
model.add(Dropout(0.2))
model.add(Dense(512, activation='relu', kernel_constraint=maxnorm(3)))
model.add(Dropout(0.2))
model.add(Dense(num_classes, activation='softmax'))

In [19]:
# Compile model
epochs = 25
lrate = 0.01
decay = lrate/epochs
sgd = SGD(lr=lrate, momentum=0.9, decay=decay, nesterov=False)
model.compile(loss='categorical_crossentropy', optimizer=sgd, metrics=['accuracy'])
print(model.summary())

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_3 (Conv2D)            (None, 32, 32, 32)        896       
_________________________________________________________________
dropout_3 (Dropout)          (None, 32, 32, 32)        0         
_________________________________________________________________
conv2d_4 (Conv2D)            (None, 32, 32, 32)        9248      
_________________________________________________________________
max_pooling2d_2 (MaxPooling2 (None, 32, 16, 16)        0         
_________________________________________________________________
conv2d_5 (Conv2D)            (None, 64, 16, 16)        18496     
_________________________________________________________________
dropout_4 (Dropout)          (None, 64, 16, 16)        0         
_________________________________________________________________
conv2d_6 (Conv2D)            (None, 64, 16, 16)        36928     
__________

In [20]:
model.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=epochs, batch_size=32,\
          callbacks=[tbCallBack])

Train on 50000 samples, validate on 10000 samples
Epoch 1/25
50000/50000 [==============================] - 13s 259us/step - loss: 1.9103 - acc: 0.2932 - val_loss: 1.5403 - val_acc: 0.4327
Epoch 2/25
50000/50000 [==============================] - 13s 258us/step - loss: 1.4194 - acc: 0.4822 - val_loss: 1.3054 - val_acc: 0.5204
Epoch 3/25
50000/50000 [==============================] - 13s 254us/step - loss: 1.1792 - acc: 0.5765 - val_loss: 1.0766 - val_acc: 0.6088
Epoch 4/25
50000/50000 [==============================] - 13s 251us/step - loss: 1.0216 - acc: 0.6365 - val_loss: 1.0025 - val_acc: 0.6462
Epoch 5/25
50000/50000 [==============================] - 13s 251us/step - loss: 0.9052 - acc: 0.6800 - val_loss: 0.8534 - val_acc: 0.6982
Epoch 6/25
50000/50000 [==============================] - 13s 259us/step - loss: 0.8270 - acc: 0.7078 - val_loss: 0.7984 - val_acc: 0.7173
Epoch 7/25
50000/50000 [==============================] - 12s 250us/step - loss: 0.7677 - acc: 0.7302 - val_loss: 0.

In [22]:
# Final evaluation of the model
scores = model.evaluate(X_test, y_test, verbose=0)
print("Accuracy: %.2f%%" % (scores[1]*100))
print("Loss: %.2f%%" % (scores[0]*100))

Accuracy: 80.23%
Loss: 60.12%


In [35]:
print(X_test.shape)
print(y_test.shape)

(10000, 3, 32, 32)
(10000, 10)


In [39]:
import pandas as pd
prediction = pd.DataFrame()
imageid = []
for i in range(len(X_test)):
    i = i + 1
    imageid.append(i)
prediction["ImageId"] = imageid
prediction["Label"] = model.predict_classes(X_test, verbose=0)
print(prediction.head())


   ImageId  Label
0        1      3
1        2      8
2        3      8
3        4      0
4        5      6


In [42]:
import numpy as np  
a  = np.array(y_test[0:5])
print('Actual labels for first five images: {0}'.format(np.argmax(a, axis=1)))

Actual labels for first five images: [3 8 8 0 6]


### Accuracy Plot from Tensorboard
![Accuracy_plot](https://github.com/Ruthvicp/CS5590_PyDL/raw/master/Module2/In_Class_Exercise/ICE4/Documentation/1.JPG)

### Loss plot from tensorboard

![Loss_plot](https://github.com/Ruthvicp/CS5590_PyDL/raw/master/Module2/In_Class_Exercise/ICE4/Documentation/2.JPG)

### Remarks
- The second model has better accuracy (82%) compared to previous model (72%)
- The trade off here is, the second model takes a bit long to train
- Predicted labels for first 5 images = [3 8 8 0 6]
- Actual labels for first 5 images = [3 8 8 0 6]